lib/dry/system/provider/source.rb

```diff
@@ -37,10 +37,20 @@ class << self
           # @see Dry::System::Provider::SourceDSL
           #
           # @api private
-          def for(name:, group: nil, &block)
-            Class.new(self) { |klass|
+          def for(name:, group: nil, superclass: nil, &block)
+            superclass ||= self
+
+            Class.new(superclass) { |klass|
               klass.source_name name
               klass.source_group group
+
+              name_with_group = group ? "#{group}->#{name}" : name
+              klass.instance_eval <<~RUBY, __FILE__, __LINE__ + 1
+                def name
+                  "#{superclass.name}[#{name_with_group}]"
+                end
+              RUBY
```

Comment:
Ruby 3.3 has `Module#set_temporary_name` https://rubyreferences.github.io/rubychanges/3.3.html#moduleset_temporary_name which would work nicely here I think

Reply:
@flash-gordon I'm all in favor of adopting language updates, but the gemspec currently allows 3.0 and up, so I'd have to do it both ways. Is there a benefit to that added complexity in order to use `set_temporary_name`?

Reply:
@alassek I have no idea :) It's up to you

Reply:
I'm happy if we just leave it for now.